xtask/src/logging.rs

```diff
@@ -1,65 +1,100 @@
-use std::path::Path;
-use std::{io::Write, path::PathBuf};
-use tracing::{info, warn};
+use std::path::{Path, PathBuf};
+use tracing::warn;
 
 // If log_path is relative it will be made absolute relative to the project_path
 //
 // The absolute path of where the log file is created is returned
-fn create_log_file(project_path: &Path, log_path: &String) -> std::io::Result<PathBuf> {
+pub fn create_log_file(project_path: &Path, log_path: &String) -> Option<String> {
     let mut log_path: PathBuf = Path::new(&log_path).to_path_buf();
 
     if log_path.is_relative() {
         log_path = project_path.join(log_path);
     }
 
     if let Some(prefix) = log_path.parent() {
-        std::fs::create_dir_all(prefix)?;
+        std::fs::create_dir_all(prefix)
+            .map_err(|e| warn!("Impossible redirect logs, using stdout instead. Error: {e}"))
+            .ok()?;
     }
-    std::fs::File::create(&log_path)?;
-    Ok(log_path)
+
+    std::fs::File::create(&log_path)
+        .map_err(|e| warn!("Impossible redirect logs, using stdout instead. Error: {e}"))
+        .ok()?;
+    Some(log_path.to_string_lossy().to_string())
 }
 
-// If the log file cannot be created due to any reasons,
-// such as lack of permission to create files or new folders in the path,
-// things will be printed to stdout instead of being redirected to the logs file
+// This macro will accept a description of the command, an xshell::Shell, a command
+// (a string on which format will be called), the log_path, and optionally the process group
+// used to create the command.
 //
-// The returned closure will accept a description of the command and the command itself as a duct::Expression.
-// The description will be printed to both stdout and the log file, if possible, while
-// to the expression will be added the redirection of the logs, if possible.
-pub fn create_with_logs(
-    project_path: &Path,
-    log_path: String,
-) -> Box<dyn Fn(&str, duct::Expression) -> duct::Expression> {
-    let without_logs = |description: &str, cmd: duct::Expression| -> duct::Expression {
-        info!("{description}");
-        cmd
-    };
-
-    let log_path = match create_log_file(project_path, &log_path) {
-        Ok(log_path) => log_path,
-        Err(e) => {
-            warn!("Impossible redirect logs, using stdout instead. Error: {e}");
-            return Box::new(without_logs);
-        }
-    };
+// The description will be logged with the info log level, and the command will be created
+// using the provided shell and converted to a tokio::process::Command to redirect stdout and stderr.
+//
+// If log is None, then things will be redirected to stdout.
+// Instead, if it contains a path, it will be used to redirect the command's output.
+//
+// The provided path needs to be a valid path, `create_log_file` should be used before use this macro
+#[macro_export]
+macro_rules! cmd_with_logs {
```

Comment:
What's the reasoning for macro? Can't we employ e.g. builder?

Reply:
The main reason I didn't implement any functions or a builder is that the conversion from `xshell::Cmd`, `std::process::Command`, and `tokio::process::Command` made things more complicated. Unable to find a perfectly clean solution, I opted for a macro to get something working quickly.

Something like the following does not work without requiring to convert things upfront

```rust
trait WithLogs {
    fn with_logs(&mut self) -> &mut Self;
    fn run_logs(&mut self) -> anyhow::Result<()>;
    fn spawn_logs(&mut self) -> anyhow::Result<tokio::process::Child>;
}
```

Reply:
Maybe I found another approach using:

```rust
trait WithLogs {
    fn with_logs(self, description: &str, log_path: Option<PathBuf>) -> tokio::process::Command;
    fn run_with_logs(self, description: &str, log_path: Option<PathBuf>) -> anyhow::Result<()>;
    fn spawn_with_logs(
        self,
        description: &str,
        log_path: Option<PathBuf>,
    ) -> anyhow::Result<tokio::process::Child>;
}

trait ProcessGroupId {
    fn process_group(self, pgid: i32) -> tokio::process::Command;
}

```

Reply:
Thank you for pushing me to rethink this! I refactored everything without needing macros and I really like how it turned out. It is now much cleaner and clearer! Hope you will think the same!!